src/main/java/org/truffleruby/language/objects/classvariables/SetClassVariableNode.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+package org.truffleruby.language.objects.classvariables;
+
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.NodeChild;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.object.FinalLocationException;
+import com.oracle.truffle.api.object.IncompatibleLocationException;
+import com.oracle.truffle.api.object.Property;
+import com.oracle.truffle.api.object.Shape;
+import org.truffleruby.core.module.ModuleOperations;
+import org.truffleruby.core.module.RubyModule;
+import org.truffleruby.language.RubyContextSourceNode;
+import org.truffleruby.language.RubyNode;
+
+@NodeChild(value = "module", type = RubyNode.class)
+@NodeChild(value = "name", type = RubyNode.class)
+@NodeChild(value = "value", type = RubyNode.class)
+public abstract class SetClassVariableNode extends RubyContextSourceNode {
+
+    public static SetClassVariableNode create() {
+        return SetClassVariableNodeGen.create(null, null, null);
+    }
+
+    public abstract Object execute(RubyModule module, String name, Object value);
+
+    @Specialization(
+            guards = {
+                    "name == cachedName",
+                    "module.getClassVariables().getShape() == cachedClassVariableStorageShape",
+                    "cachedProperty != null" })
+    protected Object setClassVariable(RubyModule module, String name, Object value,
+            @Cached("name") String cachedName,
+            @Cached("module.getClassVariables()") ClassVariableStorage cachedClassVariableStorage,
+            @Cached("cachedClassVariableStorage.getShape()") Shape cachedClassVariableStorageShape,
+            @Cached("cachedClassVariableStorage.getShape().getProperty(cachedName)") Property cachedProperty) {
+        try {
+            cachedProperty.set(cachedClassVariableStorage, value, cachedClassVariableStorageShape);
+        } catch (IncompatibleLocationException | FinalLocationException e) {
+            uncachedSetClassVariableNode(module, name, value);
+        }
```

Comment:
Could you use `@CachedLibrary("module") DynamicObjectLibrary` here instead of manually caching on the Shape?

Reply:
But that then requires looking up the name again. I know that will cache but extra things to interpret and PE when we already know the shape.

Reply:
I'm not sure to understand.
`DynamicObjectLibrary` inline caches by name (it caches on the identity of the name too) and by Shape, so I think it's the same or even better (if I read correctly, it caches on the key first, so won't even change that part of the cache).
If the Shape changes, we need to redo the lookup the anyway.
In any case, I think we should avoid manually caching on Shape or using Property as much as possible, and use the library instead.

Reply:
The problem is we need to check if the shape has the property - not just set it even if it doesn't have it anymore. I didn't want that to be two separate library calls with separate caching of the names.

It turns out `DynamicObjectLibrary` has `putIfPresent`, which does what we want. I've done that in the latest commit, but it's a couple of orders of magnitude slower - can you see what I'm doing wrong?

Reply:
`putIfPresent` seems to be slow-path? Look at `com.oracle.truffle.object.DynamicObjectLibraryImpl.SpecificKey.ExistingKey#put`.

Reply:
I think it's because you need to call putIfPresent() on the ClassVariableStorage, not on the RubyModule itself.

Reply:
Done. That worked.